src/runtime/hexagon_remote/elf.h

```diff
@@ -0,0 +1,901 @@
+                mask = 0x00df3fe0;
+            } else if ((inst >> 16) == 27209) {
+                // 0110 1010  0100 1001  PP-i iiii  i--d dddd
+                mask = 0x00001f80;
+            } else if ((inst >> 25) == 72) {
+                // 1001 0ii0  101s ssss  PPii iiii  iiid dddd
+                // 1001 0ii1  000s ssss  PPii iiii  iiid dddd
+                mask = 0x06003fe0;
+
+            } else {
+                log_printf("Unhandled!\n");
+                fail(__LINE__);
+                return;
+            }
+        }
+
+        uintptr_t old_val = val;
+        bool consumed_every_bit = false;
+        for (int i = 0; i < 32; i++) {
+            if (mask & (1 << i)) {
+                if (inst & (1 << i)) {
+                    // This bit should be zero in the unrelocated instruction
+                    fail(__LINE__);
+                    return;
+                }
+                // Consume a bit of val
+                int next_bit = val & 1;
+                if (is_signed) {
+                    consumed_every_bit |= ((intptr_t)val) == -1;
+                    val = ((intptr_t)val) >> 1;
+                } else {
+                    val = ((uintptr_t)val) >> 1;
+                }
+                consumed_every_bit |= (val == 0);
+                inst |= (next_bit << i);
+            }
+        }
+
+        if (verify && !consumed_every_bit) {
+            fail(__LINE__);
+            log_printf("Relocation overflow inst=%08lx mask=%08lx val=%08lx\n", inst, mask, (unsigned long)old_val);
+            return;
+        }
+
+        if (debug) log_printf("Relocated instruction:\n  %08lx\n", inst);
+        *((uint32_t *)addr) = inst;
+    }
+
+    // Do all the relocations for sec (e.g. .text), using the list of
+    // relocations in sec_rela (e.g. .rela.text)
+    void do_relocations_for_section(section_header_t *sec, section_header_t *sec_rela) {
+        if (!sec_rela || !sec) {
+            fail(__LINE__);
+            return;
+        }
+
+        struct known_sym {
+            const char *name;
+            char *addr;
+        };
+        static known_sym known_syms[] = {
+            {"close", (char *)(&close)},
+            {"abort", (char *)(&abort)},
+            {"memcpy", (char *)(&memcpy)},
+            {"memmove", (char *)(&memmove)},
+            {"qurt_hvx_lock", (char *)(&qurt_hvx_lock)},
+            {"qurt_hvx_unlock", (char *)(&qurt_hvx_unlock)},
+            {"__hexagon_divdf3", (char *)(&__hexagon_divdf3)},
+            {"__hexagon_muldf3", (char *)(&__hexagon_muldf3)},
+            {"__hexagon_adddf3", (char *)(&__hexagon_adddf3)},
+            {"__hexagon_divsf3", (char *)(&__hexagon_divsf3)},
+            {"__hexagon_udivdi3", (char *)(&__hexagon_udivdi3)},
+            {NULL, NULL}
+        };
+
+        // Read from the GP register for GP-relative relocations
+        char *GP;
+        asm ("{%0 = gp}\n" : "=r"(GP) : : );
```

Comment:
Perhaps move this bit to a helper function? Even though it's a one liner, it's still inline assembly...